File: api/src/main/java/com/doerapispring/domain/ListService.java
package com.doerapispring.domain;

import org.springframework.stereotype.Service;

import java.util.List;

@Service
public class ListService implements ListApplicationService {
    private final OwnedObjectRepository<TodoListCommandModel, UserId, ListId> todoListCommandModelRepository;
    private final OwnedObjectRepository<CompletedTodoList, UserId, ListId> completedTodoRepository;
    private final OwnedObjectRepository<TodoList, UserId, ListId> todoListRepository;
    private final TodoListFactory todoListFactory;
    private final ObjectRepository<User, UserId> userRepository;

    ListService(OwnedObjectRepository<TodoListCommandModel, UserId, ListId> todoListCommandModelRepository,
                OwnedObjectRepository<CompletedTodoList, UserId, ListId> completedTodoRepository,
                OwnedObjectRepository<TodoList, UserId, ListId> todoListRepository,
                TodoListFactory todoListFactory,
                ObjectRepository<User, UserId> userRepository) {
        this.todoListCommandModelRepository = todoListCommandModelRepository;
        this.completedTodoRepository = completedTodoRepository;
        this.todoListRepository = todoListRepository;
        this.todoListFactory = todoListFactory;
        this.userRepository = userRepository;
    }

    public void unlock(User user, ListId listId) {
        TodoListCommandModel todoListCommandModel = todoListCommandModelRepository.find(user.getUserId(), listId)
            .orElseThrow(ListNotFoundException::new);
        todoListCommandModel.unlock();
        todoListCommandModelRepository.save(todoListCommandModel);
    }

    public TodoListReadModel getDefault(User user) {
        return todoListCommandModelRepository.find(user.getUserId(), user.getDefaultListId())
            .map(TodoListCommandModel::read)
            .orElseThrow(ListNotFoundException::new);
    }

    @Override
    public CompletedTodoList getCompleted(User user, ListId listId) {
        return completedTodoRepository.find(user.getUserId(), listId)
            .orElseThrow(ListNotFoundException::new);
    }

    @Override
    public TodoListReadModel get(User user, ListId listId) {
        return todoListCommandModelRepository.find(user.getUserId(), listId)
            .map(TodoListCommandModel::read)
            .orElseThrow(ListNotFoundException::new);
    }

    @Override
    public List<TodoList> getAll(User user) {
        return todoListRepository.findAll(user.getUserId());
    }

    @Override
    public void create(User user, String name) {
        ListId listId = todoListRepository.nextIdentifier();
        TodoList todoList = todoListFactory.todoList(user.getUserId(), listId, name);
        todoListRepository.save(todoList);
    }

    @Override
    public void setDefault(User user, ListId listId) {
        userRepository.save(new User(user.getUserId(), listId));
    }
}

File: api/src/main/scala/com/doerapispring/domain/ListService.scala
package com.doerapispring.domain

import java.time.Clock
import java.util
import java.util.Date

import com.doerapispring.domain.events.TodoListEvent
import org.springframework.stereotype.Service
import scala.jdk.CollectionConverters._

@Service
class ListService(val todoListRepository: OwnedObjectRepository[TodoList, UserId, ListId],
                  val todoListFactory: TodoListFactory,
                  val userRepository: ObjectRepository[User, UserId],
                  val todoListModelRepository: OwnedObjectReadRepository[TodoListModel, UserId, ListId],
                  val clock: Clock,
                  val domainEventPublisher: DomainEventPublisher[TodoListModel, TodoListEvent, UserId, ListId],
                  val todoListEventRepository: OwnedObjectWriteRepository[TodoListEvent, UserId, ListId],
                  val todoListModelSnapshotRepository: OwnedObjectWriteRepository[Snapshot[DeprecatedTodoListModel], UserId, ListId])
  extends ListApplicationService {

  override def performOperation(user: User,
                                listId: ListId,
                                event: TodoListEvent): TodoListModel = {
    todoListModelRepository.find(user.getUserId, listId)
      .map(todoList => TodoListModel.applyEvent(todoList, event))
      .map(todoList => domainEventPublisher.publish(todoList, event, user.getUserId, listId))
      .get
  }

  override def getDefault(user: User): TodoListModel = {
    todoListModelRepository.find(user.getUserId, user.getDefaultListId).get
  }

  override def getCompleted(user: User, listId: ListId): List[CompletedTodo] = {
    todoListModelRepository.find(user.getUserId, listId).get.completedTodos
  }

  override def get(user: User, listId: ListId): TodoListModel = {
    todoListModelRepository.find(user.getUserId, listId).get
  }

  override def getAll(user: User): List[TodoList] = {
    todoListRepository.findAll(user.getUserId).asScala.toList
  }

  override def create(user: User, name: String): Unit = {
    val listId = todoListRepository.nextIdentifier
    val todoList = todoListFactory.todoList(user.getUserId, listId, name)
    todoListRepository.save(todoList)
    todoListModelSnapshotRepository.save(
      user.getUserId,
      listId,
      Snapshot(DeprecatedTodoListModel(listId, name, List(), new Date(0L), 0), Date.from(clock.instant())))
  }

  override def setDefault(user: User, listId: ListId): Unit = {
    userRepository.save(new User(user.getUserId, listId))
  }
}
